EVG-6620 support EBS volumes for spawn hosts

ablack12 opened the pull request:
Most of what I did was write tests, fix bugs, switch to using env and serializing into data structures and not directly into the handlers for the API routes Thomas wrote. (His original PR: https://github.com/evergreen-ci/evergreen/pull/2767)

The new commands are:
```
evergreen volume create --type TYPE --size SIZE --zone AVAILABILITY_ZONE
evergreen volume delete --id VOLUME_ID
evergreen host attach-volume --host HOST_ID  --volume VOLUME_ID --device DEVICE_NAME
evergreen host detach-volume --host HOST_ID --volume VOLUME_ID
```

Sorry like four commits are merges l o l

tychoish commented on model/host/db.go:
```diff
@@ -998,6 +997,52 @@ func AggregateLastContainerFinishTimes() ([]FinishTime, error) {
 
 }
 
+func AddVolumeToHost(h *Host, newVolume VolumeAttachment) error {
```
I think you can do these both as methods on *Host, and you just pass the pointer reciever to the findAndModify... 

I think the current api implications of making this a function (and the fact that it modifies the host you pass it anyway, is confusing.

ablack12 replied:
I'm good to make it a method, I just didn't see any others in this file. I still think this should be in db.go given the FindAndModify though.

john-m-liu commented on cloud/ec2.go:
```diff
@@ -1059,6 +1063,18 @@ func (m *ec2Manager) OnUp(ctx context.Context, h *host.Host) error {
 	defer m.client.Close()
 
+	// Add instance stores and volumes to host document
+	devices, err := m.client.GetInstanceBlockDevices(ctx, h)
```
Instead of doing this, can we also save the block device mappings when we get the DNS name? It looks like they are return in the same API call

john-m-liu replied:
Looks like Thomas already thought of this. He saved it when GetInstanceStatus is called, so I think you can remove the API call here

john-m-liu commented on model/host/db.go:
```diff
@@ -21,7 +21,8 @@ import (
 	// Collection is the name of the MongoDB collection that stores hosts.
-	Collection = "hosts"
+	Collection        = "hosts"
+	VolumesCollection = "volumes"
```
Do we need a new collection for these? Would it work to embed them in the host doc?

ablack12 replied:
This was more a conversation when Thomas first started this, but yes we need a new collection because volumes are created independent of hosts and then later detached, so they can't be tied to the host collection.

john-m-liu commented on model/host/volume.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+// Remove a volume from the volumes collection.
+func (v *Volume) Remove() error {
```
Unless you ensure that no hosts are referencing this, I don't think you can remove the document entirely - it'll need to be a flag that it's been deleted in AWS

john-m-liu replied:
It's called from https://github.com/evergreen-ci/evergreen/pull/2805/files/c3dfc87d54933a724d9955b592cc54946d84cac5#diff-e5707ed36e7443139f523d8951617a72R1149

ablack12 replied:
Oh touche. But [theoretically](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-deleting-volume.html) we can't delete a volume if it's attached to a host anyway, so I think this should be okay.

john-m-liu commented on rest/route/host_spawn.go:
```diff
@@ -227,37 +205,49 @@ func checkInstanceTypeHostStopped(h *host.Host) error {
 	return nil
 }
 
+func checkExpirableHostLimitExceeded(userId string) error {
```
Just double checking - did you mean to include changes to forever hosts here?

ablack12 replied:
Yeah sorry to mix things--because I moved options into something the user passes in the body, this allows you to extend the expiration using this host modify (which seems fine to me) but then this should also check expirable limits.

john-m-liu commented on units/spawnhost_attach_volume.go:
```diff
@@ -0,0 +1,101 @@
+package units
```
Do these need to be jobs? If they don't take too long, I think it would be easier to have our APIs call AWS APIs directly so that the caller knows when it's done

ablack12 commented:
@tychoish this conversation was so long ago I just want to make sure, was there an actual argument for having create/delete volumes be separate from the host right? Or else CreateVolume/AttachVolume (and vise versa) could really be combined.

tychoish commented:
I think it has to be seperate, because people are going to want to move their volumes from one host to another.

ablack12 commented:
Some updates:
* I removed the jobs because actually these calls don't take long unless there's a problem, and I think we can avoid most problems with more validation.
* I added an AvailabilityZone flag (and I added this information to host.list) so users can more easily create a volume in their hosts' zone.

ablack12 commented:
The failing tests are a John bug and the cloud setup-failure we've been seeing pretty consistently so I'm just going to merge.
